pytest_bdd/scenario.py

```diff
@@ -103,7 +104,10 @@ def _execute_step_function(request, scenario, step, step_func):
         request.config.hook.pytest_bdd_before_step_call(**kw)
         target_fixture = getattr(step_func, "target_fixture", None)
         # Execute the step.
-        return_value = step_func(**kwargs)
+        if sync:
+            return_value = step_func(**kwargs)
+        else:
+            return_value = await step_func(**kwargs)
```

Comment:
so you're saying?
```python
        if (force_async or iscoroutinefunction(step_func)) and not sync:
            return_value = await step_func(**kwargs)
        else:
            return_value = step_func(**kwargs)
```

this way Twisted users would write:


```python
@given(force_async=True)
def has_something():
    return treq.get("https://example.com")
```

Reply:
or something like this?

```python
def force_async(fn):
    async def wrapper(*args, **kwargs):
         return await fn(*args, **kwargs)

    return wrapper

def force_sync(fn):
    async def wrapper(*args, **kwargs:
        return fn(*args, **kwargs)
    return wrapper


scenarios(..., sync=False)

@given(...)
@force_async
def has_something():
    return treq.get("https://example.com")

@given(..., target_fixture=the_given)
@force_sync
def the_given():
 return real_setup()

@when(...)
async def(the_given):
 await act_on(the_given)
```

Reply:
Yeah, something like one of those. I think I prefer the former, but obviously haven't used them in anger.

My underlying assumption is that if you have a coroutinefunction, you are very likely to expect to await it, e.g. in trio you _must_ await it; with asyncio you _should_ await it, unless you are actually trying to return the "awaitable" to use later with the requisite risk of accidentally _never_ awaiting it. I think the "unless" version there is rare enough that flagging it explicitly in the step definition is fine; it's also probably a particularly odd thing to do in a BDD-style test.

If it's not a coroutinefunction, I'd suggest that it's more commonly just a plain function to call (i.e. code that is working in pytest-bdd today; I'd make `@force_sync` the default behaviour there), and making it possible to say "This non-coroutine function is returning an awaitable to be awaited" covers the `treq.get` case there.

Could the treq.get case be simplified as
```
@given(...)
async def has_something():
    return await treq.get("https://example.com")
```
so you don't need to `@force_async` there? This `step_func` would be awaited by default (being `async def`), and that will await `treq.get` explicitly in the step, where we know it's the right thing to do, assuming we know that `treq.get`'s return value needs to be awaited.

That feels more natural to me, but I haven't used `treq` though, so I might be overlooking something there.

Reply:
Right, but Twisted projects are the opposite of Trio, in that await is currently rarely used and .addCallbacks is used instead

Reply:
That makes sense, thank you for clarifying. I used Twisted heavily almost a decade ago, i.e. on Python 2, before `await`; so wasn't sure if the common usage patterns had shifted towards `await` in the meantime.